File: documentation-examples-scala/src/scala/dev/typr/foundationssc/docs/core/QueryListenerSlowQuery.scala
package dev.typr.foundationssc.docs.core
import dev.typr.foundationssc.*
import dev.typr.foundationssc.data.*

import dev.typr.foundations.{QueryListener, QueryEvent}
import java.time.Duration

@SuppressWarnings(Array("unused"))
object QueryListenerSlowQuery:
  //start
  def slowQueryDetector(threshold: Duration): QueryListener =
    new QueryListener:
      override def beforeQuery(sql: String, name: String): Unit = ()
      override def afterQuery(event: QueryEvent): Unit =
        if event.elapsed().compareTo(threshold) > 0 then
          System.err.println(s"SLOW QUERY [${event.elapsed().toMillis}ms]: ${event.interpolatedSql()}")
      override def failedQuery(event: QueryEvent): Unit = ()
  //stop

File: documentation-examples-java/src/java/dev/typr/foundations/docs/core/QueryListenerSlowQuery.java
package dev.typr.foundations.docs.core;

import dev.typr.foundations.QueryEvent;
import dev.typr.foundations.QueryListener;
import java.time.Duration;

@SuppressWarnings("unused")
public class QueryListenerSlowQuery {
  // start
  QueryListener slowQueryDetector(Duration threshold) {
    return new QueryListener() {
      @Override
      public void beforeQuery(String sql, String name) {}

      @Override
      public void afterQuery(QueryEvent event) {
        if (event.elapsed().compareTo(threshold) > 0) {
          System.err.println(
              "SLOW QUERY [" + event.elapsed().toMillis() + "ms]: " + event.interpolatedSql());
        }
      }

      @Override
      public void failedQuery(QueryEvent event) {}
    };
  }
  // stop
}
